compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -212,29 +222,335 @@ public void visit(BLangPackage pkgNode) {
+        
+        initFunction.desugared = false;
+        initFunction.requiredParams.add(blockClosureMap);
+        SymbolEnv env = oceData.objMethodsEnv;
+        
+        BVarSymbol paramSym = new BVarSymbol(Flags.FINAL, mapSymbol.name, env.scope.owner.pkgID, mapSymbol.type,
+                initFunction.symbol, classDef.pos, VIRTUAL);
+
+        // update symbols for init method
+        initFunction.symbol.scope.define(paramSym.name, paramSym);
+        initFunction.symbol.params.add(paramSym);
+        BInvokableType initFuncSymbolType = (BInvokableType) initFunction.symbol.type;
+        initFuncSymbolType.paramTypes.add(mapSymbol.type);
+
+        BInvokableType initFnType = (BInvokableType) initFunction.getBType();
+        initFnType.paramTypes.add(mapSymbol.type);
+
+        BAttachedFunction attachedFunction = ((BObjectTypeSymbol) classDef.getBType().tsymbol).generatedInitializerFunc;
+        attachedFunction.symbol.params.add(paramSym);
+        classDef.generatedInitFunction = desugar.rewrite(classDef.generatedInitFunction, oceData.objMethodsEnv);
+
+        // Add map to the callee expression
+        addMapToCalleeExpression(mapSymbol, oceData);
+    }
+
+    private void addMapToCalleeExpression(BVarSymbol mapSymbol, OCEDynamicEnvironmentData oceData) {
+        BLangSimpleVarRef.BLangLocalVarRef blockLevelMapLocalVarRef = new BLangSimpleVarRef.BLangLocalVarRef(mapSymbol);
+        oceData.attachedFunctionInvocation.requiredArgs.add(blockLevelMapLocalVarRef);
+        visit(oceData.attachedFunctionInvocation);
+    }
+
+    private void addMapSymbolAsAField(BLangClassDefinition classDef, BVarSymbol mapSymbol) {
+        BLangSimpleVariable mapVar = ASTBuilderUtil.createVariable(symTable.builtinPos, mapSymbol.name.value,
+                mapSymbol.type, null, mapSymbol);
+
+        BObjectType object = (BObjectType) classDef.getBType();
+        object.fields.put(mapSymbol.name.value, new BField(mapSymbol.name, classDef.pos, mapSymbol));
+        classDef.fields.add(0, mapVar);
+    }
+
+    private void addFunctionLevelClosureMapToClassDefinition(BLangClassDefinition classDef) {
+        if (classDef.oceEnvData.functionMapUpdatedInInitMethod || classDef.oceEnvData.mapFunctionMapSymbol == null) {
+            return;
+        }
+
+        BLangFunction function = (BLangFunction) classDef.oceEnvData.capturedClosureEnv.enclInvokable;
+        BVarSymbol functionMap = createMapSymbolIfAbsent(function, blockClosureMapCount);
+        BVarSymbol mapSymbol = classDef.oceEnvData.mapFunctionMapSymbol;
+
+        addMapSymbolAsAField(classDef, mapSymbol);
+        createClosureMapUpdateExpression(classDef, functionMap, mapSymbol);
+        addClosureMapToInit(classDef, functionMap);
+
+        classDef.oceEnvData.functionMapUpdatedInInitMethod = true;
+    }
+
+    /**
+     * This function adds block level closure map to class as a field and as an argument to init. The object
```

Comment:
Shall we add a code example to get an idea what we desugar here? original source and the desugared code